Tag filter updates

fsnoddy opened the pull request:
…(Can be useful for AA tracking)

Add an "Or" option for different groupings
Modify Filter max width to be in pixels so that it can be nicely nested in a column

**Related issues / Requêtes associées**
Reopen #9796

fsnoddy committed:
- Add option to not filter live on the Filter and Tag Filter plugins (Can be useful for AA tracking)
- Add an "Or" option for different groupings
- Modify Filter max width to be in pixels so that it can be nicely nested in a column

fsnoddy committed:
- Cleanup PR
- Fix french example
- Remove testing code

fsnoddy committed:
Fix eslint issues.

fsnoddy committed:
Aria not supported on <form> element

fsnoddy committed:
Very minor change, fixed "that" to "self" as a best practice. I believe I had seen it in another plugin which is why this nomenclature had been used.

fsnoddy committed:
Exception added for search element to address linting issue until the Nu validator's issue is fixed.

fsnoddy committed:
Aria label not necessary since the heading already adds context to the element.

fsnoddy committed:
Lint issues

fsnoddy committed:
Update tagfilter.js

fsnoddy committed:
Update Gruntfile.coffee

Co-authored-by: EricDunsworth <1907279+EricDunsworth@users.noreply.github.com>

fsnoddy committed:
Update src/plugins/tagfilter/tagfilter-fr.hbs

Co-authored-by: EricDunsworth <1907279+EricDunsworth@users.noreply.github.com>

fsnoddy committed:
Update src/plugins/tagfilter/tagfilter-fr.hbs

Co-authored-by: EricDunsworth <1907279+EricDunsworth@users.noreply.github.com>

fsnoddy committed:
Update site/pages/docs/ref/filter/filter-en.hbs

Co-authored-by: Garneauma <garneauma@gmail.com>

fsnoddy committed:
- Changed the french documentation to english as requested.
- Added group option to fieldset
- Added tag filter configurations options
- Added example for the Filter plugin

fsnoddy committed:
Resolving the merge conflicts introduced linting issues

fsnoddy commented on src/plugins/tagfilter/tagfilter.js:
```diff
@@ -213,6 +201,22 @@ const componentName = "wb-tagfilter",
 	},
 
+	getControlFilterName = function( control ) {
+		const controlLegend = $( control ).closest( "fieldset[data-" + componentName + "]" );
```
Sure thing, that would make a lot of sense.

fsnoddy commented on src/plugins/tagfilter/tagfilter.js:
```diff
@@ -213,6 +201,22 @@ const componentName = "wb-tagfilter",
+		const controlLegend = $( control ).closest( "fieldset[data-" + componentName + "]" );
+		let controlSettings = {};
+
+		if ( control.getAttribute( "data-" + componentName ) !== null ) {
```
Ah params go only on the component itself and not children. Got it, sorry for the confusion

fsnoddy commented on src/plugins/tagfilter/tagfilter-fr.hbs:
```diff
@@ -1000,6 +1001,254 @@
+
+<details>
+	<summary>Code source</summary>
+	<pre><code>&lt;section id="filteredFinder" class="wb-tagfilter provisional wb-filter" data-wb-tagfilter='{ "live": false }' data-wb-filter='{
```
@Garneauma You don't want a demo of the "live" feature at all or need a separate example?

Garneauma replied:
@fsnoddy My mistake. Please disregard this comment.

fsnoddy committed:
Made the updates to use a different data attribute than the component name for the groups

fsnoddy committed:
Update src/i18n/i18n.csv

Co-authored-by: Garneauma <garneauma@gmail.com>

fsnoddy committed:
Update src/plugins/filter/filter-fr.hbs

Co-authored-by: Garneauma <garneauma@gmail.com>

fsnoddy committed:
Update src/plugins/filter/filter-fr.hbs

Co-authored-by: Garneauma <garneauma@gmail.com>

fsnoddy committed:
Reviewed and updated documentation

fsnoddy committed:
Filter change to pick up form submits from the UiTemplate

fsnoddy commented on site/pages/docs/ref/filter/filter-en.hbs:
```diff
@@ -57,6 +57,15 @@
 	...
 &lt;/ul&gt;</code></pre>
+		</li>
+		<li><p><strong>Optional:</strong> Set the live option to false and include a form element in the container.</p>
```
It's a very good point, I've amended the documentation to make that less specific on one specific scenario.

- If UiTemplate is NOT defined, the live attribute adds a button automatically to the controls prepended to the component.
- If UiTemplate is defined, the event handler will pick up a form submit.

Garneauma reviewed (changes requested):
@fsnoddy I'm sorry for all the partial reviews.

- Preferably, the items should be updated upon hitting "ENTER" when the "live" option is set to "false". This would mean that the field and button would need to be wrapped in a form element.
- Please add an instruction in the Tagfilter plugin for when using the "live: false" option that users need to add a form element around the filters.

fsnoddy committed:
- Ajouté le support pour la touche enter dans le plugiciel filter
- Mise-à-jour la documentation

fsnoddy commented on site/pages/docs/ref/filter/filter-en.hbs:
```diff
@@ -251,6 +271,7 @@
 	<h3 id="uiTemplateEB">uiTemplate</h3>
 	<ul>
 		<li>The template <strong>must</strong> include an <code>&lt;input type="search"></code>.</li>
+		<li>The template <strong>must</strong> include a <code>&lt;button></code> if the <code>live: false</code> option is used.</li>
```
I've added support the enter key.

Garneauma replied:
@fsnoddy I think the instructions need to be changed for when using a custom UI. I just tried to set the option "live" to false, but it wouldn't work unless the input and button were wrapped by a `form` element.

fsnoddy committed:
Merge branch 'master' into tagFilterUpdates

fsnoddy committed:
Fix lint issue

fsnoddy committed:
There was a class missing to enable the event. The test should no work.

EricDunsworth commented on src/plugins/tagfilter/tagfilter-en.hbs:
```diff
@@ -981,6 +983,252 @@
+		"uiTemplate": "#searchFilter2",
+		"live": false
+	}'>
+	<h2>Example 4 - Finder with "or" behavior updating with a button</h2>
```
Hmm actually... maybe the opposite needs to be done for this (i.e. updating the ToC link and code sample to match this heading).

The French version of this content is currently called "Exemple 4 - Chercheur de types "ou" mis-à-jour avec bouton" - which is in line with the English heading.

EricDunsworth replied:
Ended up "normalizing" those ToC links + headings to these:
* **English:** Example 4 - Finder with "or" behavior and submit button
* **Français:** Exemple 4 - Chercheur de type «&nbsp;ou&nbsp;» et bouton de soumission

EricDunsworth committed:
Merge branch 'master' into tagFilterUpdates

Conflicts:
	src/plugins/filter/filter-en.hbs
	src/plugins/filter/filter-fr.hbs
	src/plugins/filter/filter.js
	src/plugins/tagfilter/tagfilter.js

EricDunsworth committed:
Content: Collapse groups of empty lines into one

EricDunsworth committed:
Content: Increase new filter demo's heading levels

EricDunsworth committed:
Content: Normalize new demos' links/headings

They were previously inconsistently-named...

EricDunsworth committed:
Content: Update new demo's code samples

EricDunsworth committed:
Filter plugin: Let the tag filter intercept form submissions

The new tag filter demo previously only partially-worked. Submitting the filters form only honoured the text field (powered by the filter plugin). Checkbox and dropdown selections (powered by the tag filter plugin) were completely ignored. Why? Because the filter plugin's preventDefault() method prevented the tag filter plugin from intercepting the form submit event.

This resolves it by preventing filter plugin's call to preventDefault() from running if the filter and tag filter plugins' classes (wb-filter and wb-tagfilter) are being used together.

Co-authored-by: fsnoddy <francis.snoddy@hrsdc-rhdcc.gc.ca>

EricDunsworth committed:
Plugins: Space-out lines in new logic

* Matches the tag filter plugin's pre-existing spacing
* Filter plugin doesn't appear to use consistent spacing
  * Note: Decided against reworking pre-existing logic's spacing as a part of these changes

EricDunsworth commented on Gruntfile.coffee:
```diff
@@ -892,6 +892,9 @@ module.exports = (grunt) ->
 						"The “contentinfo” role is unnecessary for element “footer”."
 						"The “navigation” role is unnecessary for element “nav”."
 						"The “banner” role is unnecessary for element “header”."
+						"Element “search” not allowed as child of element “main” in this context. (Suppressing further errors from this subtree.)"
+						"Element “search” not allowed as child of element “section” in this context. (Suppressing further errors from this subtree.)"
+						"Element “search” not allowed as child of element “div” in this context. (Suppressing further errors from this subtree.)"
```
FWIW the HTML validator officially added support for the ``<search>`` element in early June via validator/validator#1828, but... but there haven't been any "proper" new releases of the validator since then ([latest tag](https://github.com/validator/validator/tags) is called ``latest``... not ideal since it's at risk of changing constantly).

If a new version number tag gets rolled-out at some point, hopefully ``grunt-html`` will quickly receive a new release as well ([its latest release](https://github.com/validator/validator/tags) came out right after the validator's latest numbered tag).

Maybe it'd be possible to override WET's ``grunt-html`` dependency to use a newer validator commit hash in the meantime...

EricDunsworth committed:
Merge branch 'master' into tagFilterUpdates

Conflicts:
	site/pages/docs/ref/tagfilter/tagfilter-en.hbs
	site/pages/docs/ref/tagfilter/tagfilter-fr.hbs
	src/plugins/tagfilter/tagfilter-en.hbs
	src/plugins/tagfilter/tagfilter-fr.hbs
	src/plugins/tagfilter/tagfilter.js

EricDunsworth committed:
Content: Remove misc empty lines

EricDunsworth committed:
Content: Rename date range demo to example 6

EricDunsworth committed:
Content: Update modified dates

Garneauma commented:
@fsnoddy @EricDunsworth Do you still intend on finishing this work in the near future? If not, can this be transitioned to a draft PR? Thanks!

EricDunsworth commented:
@Garneauma Sorry :( - just noticed your comment.

I intend to try getting this across the finish line at some point...
* Proper support for radio buttons (i.e. "OR" option) is an obvious use case for something like the tag filter plugin
* We have a need for its functionality in the [life events' checklist pages](https://github.com/ServiceCanada/canada-life-events/blob/72211981061d198acb0875db181c4096b7fa0f5f/pages/th-retirement-hub/templates/checklist.html) (currently using [custom JS logic](https://github.com/ServiceCanada/canada-life-events/blob/72211981061d198acb0875db181c4096b7fa0f5f/src/life-events.js#L3-L97) as a stop-gap measure)
